### PR TITLE
Integrate DOMPurify

### PR DESCRIPTION
**Pull request by choumx**
### Summary

Partial for #15123. ~Blocked by #11115 [1].~

- Replaces Caja with DOMPurify under `svg-in-mustache` experiment.

May break pages if there are cases where DOMPurify is more strict than Caja. It seems less strict that Caja overall though since attribute whitelists are not tag-specific, etc.

[1] DOMPurify eagerly removes attributes and then re-adds them after sanitization, which doesn't work for binding attributes like `[foo]` so we need to rewrite them to use the `data-amp-bind-foo` syntax.

### Size impact

Adds 5776B to `amp-mustache.js` (gzipped, minified). Looks about right since DOMPurify itself is 5707B minified. However, we'll save ~3kB once we remove the experiment and clean up the Caja integration.

| Caja (today) | Caja + DOMPurify | DOMPurify |
| -- | -- | -- |
| 15070 | 20846 | 12004 |

### CODE REVIEW

**Comment by choumx** [comment]
Adds 5776B to `amp-mustache.js` (gzipped, minified). Looks about right since DOMPurify itself is 5707B minified. However, we'll save ~3kB once we remove the experiment and clean up the Caja integration.

| Caja (today) | Caja + DOMPurify | DOMPurify |
| -- | -- | -- |
| 15070 | 20846 | 12004 |

**Comment by choumx** [comment]
@molnarg PTAL, would appreciate your feedback.

File: src/sanitizer.js [hunk 1]
```diff
@@ -213,6 +248,98 @@ const INVALID_INLINE_STYLE_REGEX =
...
+    }
+  }
+}
+
+/**
+ * @param {!Node} node
+ * @param {{attrName: string, attrValue: string, allowedAttributes: !Array<string>}} data
+ */
+function uponSanitizeAttribute(node, data) {
+  const tagName = (node.tagName || '').toLowerCase();
```
Comment (molnarg) [inline src/sanitizer.js]: I was about to warn here about DOM clobbering, but it turns out DOMPurify just drops clobbered DOM elements whose tagName is unreliable \o/ https://github.com/cure53/DOMPurify/blob/master/src/purify.js#L527 It'll also filter elements with some other clobbered properties. Please check if you only use the properties that are guaranteed to be unclobbered (I've quickly ran through them and didn't see anything you use that is not protected). It's also worth a comment for future updates to this function.

File: src/sanitizer.js [hunk 2]
```diff
@@ -129,13 +139,18 @@ const WHITELISTED_ATTRS = [
...
   'subscriptions-service',
-  'subscriptions-decorate',
 ];
 
-/** @const {!Object<string, !Array<string>>} */
+/**
+ * Tag-specific attribute whitelist for Caja. DOMPurify uses a tag-agnostic
+ * whitelist and custom attrs below are duped into PURIFY_CONFIG['ADD_ATTR'].
+ * @const {!Object<string, !Array<string>>}
+ */
```
Comment (molnarg) [inline src/sanitizer.js]: Some of these are valid in other contexts as well (e.g. rel) so I'd rather keep them as specific as possible. The tag specificity could be kept by whitelisting these in the attribute hook after checking the tagName.

**Comment by choumx** [comment]
> It'd be nice to have some tests, at least for the newly added SVG support. 

Done.

> Also, could you run it against the DOMPurify test suite to see if we have anything bad whitelisted in our config?

These are the full results: https://gist.github.com/choumx/85795aec306fb5e4cabdd60c937ee8d5

Most of the breakages are due to blacklisting and custom modificaitons e.g. adding `target` attribute for `a` tags, rewriting relative URLs to be absolute. Commenting out those customizations while leaving the blacklists only results in two test failures.

```sh
willchou$ git diff src/sanitizer.js
diff --git a/src/sanitizer.js b/src/sanitizer.js
index c2218254c..19d0f8210 100644
--- a/src/sanitizer.js
+++ b/src/sanitizer.js
@@ -257,8 +257,8 @@ export function sanitizeHtml(html) {
 function purifyHtml(dirty) {
   const config = Object.assign({}, PURIFY_CONFIG, {
     'ADD_ATTR': WHITELISTED_ATTRS,
-    'FORBID_ATTR': isExperimentOn(self, 'inline-styles') ? [] : ['style'],
-    'FORBID_TAGS': Object.keys(BLACKLISTED_TAGS),
+    // 'FORBID_ATTR': isExperimentOn(self, 'inline-styles') ? [] : ['style'],
+    // 'FORBID_TAGS': Object.keys(BLACKLISTED_TAGS),
   });
   DOMPurify.addHook('uponSanitizeElement', uponSanitizeElement);
   DOMPurify.addHook('uponSanitizeAttribute', uponSanitizeAttribute);
@@ -277,11 +277,11 @@ function uponSanitizeElement(node, data) {
     allowedTags[tagName] = true;
   }
 
-  if (tagName == 'a') {
-    if (node.hasAttribute('href') && !node.hasAttribute('target')) {
-      node.setAttribute('target', '_top');
-    }
-  }
+  // if (tagName == 'a') {
+  //   if (node.hasAttribute('href') && !node.hasAttribute('target')) {
+  //     node.setAttribute('target', '_top');
+  //   }
+  // }
 }
 
 /**
@@ -330,9 +330,9 @@ function uponSanitizeAttribute(node, data) {
   }
 
   if (isValidAttr(tagName, attrName, attrValue, /* opt_purify */ true)) {
-    if (attrValue && !startsWith(attrName, 'data-amp-bind-')) {
-      attrValue = rewriteAttributeValue(tagName, attrName, attrValue);
-    }
+    // if (attrValue && !startsWith(attrName, 'data-amp-bind-')) {
+    //   attrValue = rewriteAttributeValue(tagName, attrName, attrValue);
+    // }
   } else {
     user().error(TAG, `Removing "${attrName}" attribute with invalid `
         + `value in <${tagName} ${attrName}="${attrValue}">.`);
@@ -534,12 +534,12 @@ export function isValidAttr(tagName, attrName, attrValue, opt_purify = false) {
   }
 
   // Inline styles are not allowed.
-  if (attrName == 'style') {
-    if (isExperimentOn(self, 'inline-styles')) {
-      return !INVALID_INLINE_STYLE_REGEX.test(attrValue);
-    }
-    return false;
-  }
+  // if (attrName == 'style') {
+  //   if (isExperimentOn(self, 'inline-styles')) {
+  //     return !INVALID_INLINE_STYLE_REGEX.test(attrValue);
+  //   }
+  //   return false;
+  // }
 
   // Don't allow CSS class names with internal AMP prefix.
   // See https://github.com/ampproject/amphtml/blob/master/validator/validator-main.protoascii
```

```
SUMMARY:
● 188 tests completed
● 2 tests failed

FAILED TESTS:
  DOMPurify-based
    DOMPurify test suite
      ● MathML example
        Chrome 66.0.3359 (Mac OS X 10.13.4)
      AssertionError: expected [ Array(3) ] to include ''
          at Context.<anonymous> (/var/folders/xy/zdjxmyvn1vjg75zhtgc5r28h006wch/T/test/functional/test-sanitizer.js:156:30 <- /var/folders/xy/zdjxmyvn1vjg75zhtgc5r28h006wch/T/0c70690797a399b84c6fdbae6991b94f.browserify:75254:41)

      ● MathML
        Chrome 66.0.3359 (Mac OS X 10.13.4)
      AssertionError: expected [ Array(2) ] to include '<div id="130">\n//["\'`--&gt;]]&gt;]</div>'
          at Context.<anonymous> (/var/folders/xy/zdjxmyvn1vjg75zhtgc5r28h006wch/T/test/functional/test-sanitizer.js:156:30 <- /var/folders/xy/zdjxmyvn1vjg75zhtgc5r28h006wch/T/0c70690797a399b84c6fdbae6991b94f.browserify:75254:41)
```
So looks good overall I think.

**Comment by molnarg** [comment]
Thanks. And good catch for tagName vs. nodeName, I though the one the code was using is the correct one!

**Pull request merged**